File: frontend/src/types/report.ts
import { z } from "zod";

export type ReportType =
  | "overview"
  | "asset_inventory"
  | "device_health"
  | "onboarding_summary"
  | "compliance_audit";

export type JobStatus = "queued" | "running" | "completed" | "failed";

// Matches ReportJobResponse from backend/schemas/report.py
export interface ReportJobResponse {
  job_id: string;
  status: JobStatus;
  error_message?: string | null;
}

// Matches ReportRead from backend/schemas/report.py
export interface ReportRead {
  id: string;
  report_name: string;
  report_type: string;
  triggered_by: string;
  recipient_emails: string;
  status: string;
  sent_at: string;
  error_message?: string | null;
}

export const reportTypeLabels: Record<ReportType, string> = {
  overview: "System Overview Report",
  asset_inventory: "Asset Inventory Report",
  device_health: "Device Health Audit",
  onboarding_summary: "Onboarding Workflow Summary",
  compliance_audit: "Compliance & Security Audit Log",
};

export const reportTriggerSchema = z
  .object({
    report_type: z.enum(
      [
        "overview",
        "asset_inventory",
        "device_health",
        "onboarding_summary",
        "compliance_audit",
      ],
      {
        required_error: "Bitte wählen Sie einen Berichtstyp aus.",
      }
    ),
    start_date: z.string().optional(),
    end_date: z.string().optional(),
  })
  .refine(
    (data) => {
      if (data.start_date && data.end_date) {
        return new Date(data.start_date) <= new Date(data.end_date);
      }
      return true;
    },
    {
      message: "Das Startdatum muss vor dem Enddatum liegen.",
      path: ["start_date"],
    }
  );

export type ReportTriggerInput = z.infer<typeof reportTriggerSchema>;

File: backend/schemas/report.py
import uuid
from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class ReportType(str, Enum):
    SCHEDULED_WEEKLY = "scheduled_weekly"
    MANUAL = "manual"
    ALERT = "alert"


class ReportTriggerRequest(BaseModel):
    report_type: Literal[
        "overview", "asset_inventory", "device_health", "onboarding_summary", "compliance_audit"
    ] = Field(..., description="The classification of the report to generate.")
    start_date: datetime | None = Field(None, description="Start date for the report data range.")
    end_date: datetime | None = Field(None, description="End date for the report data range.")


class ReportJobResponse(BaseModel):
    job_id: uuid.UUID
    status: Literal["queued", "running", "completed", "failed"]
    error_message: str | None = None


class ReportRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    report_name: str
    report_type: str
    triggered_by: str
    recipient_email: str
    status: str
    sent_at: datetime
    error_message: str | None = None
